stellargraph/connector/neo4j/graph.py

```diff
@@ -0,0 +1,113 @@
+__all__ = ["Neo4jStellarGraph"]
+
+import numpy as np
+import scipy.sparse as sps
+from ...core.experimental import experimental
+
+
+@experimental(reason="the class is not fully tested and lacks documentation")
+class Neo4jStellarGraph:
+    def __init__(self, graph_db, is_directed=False):
+
+        self.graph_db = graph_db
+        self._is_directed = is_directed
+
+    def nodes(self):
+        # FIXME: don't assume stellargraphs "ID"s are in the Neo4j graph
+        # use id() instead
+        node_ids_query = f"""    
+            CALL apoc.cypher.run(
+                'MATCH (n)
+                RETURN n.ID as node_ids',
+                {{}}
+            ) YIELD value
+
+            RETURN collect(value.node_ids) as node_ids
+            """
+        result = self.graph_db.run(node_ids_query)
+
+        return result.data()[0]["node_ids"]
+
+    def node_features(self, node_ids):
+        feature_query = f"""
+            UNWIND $node_id_list AS node_id
+
+            // for each node id in every row, collect the random list of its neighbors.
+            CALL apoc.cypher.run(
+
+                'MATCH(cur_node) WHERE cur_node.ID = $node_id
+                RETURN cur_node.features as features',
+                {{node_id: node_id}}
+            ) YIELD value
+
+            RETURN collect(value.features) as features
+            """
+
+        result = self.graph_db.run(
+            feature_query, parameters={"node_id_list": node_ids},
+        )
+
+        return np.array(result.data()[0]["features"])
+
+    def to_adjacency_matrix(self, node_ids):
+        subgraph_query = f"""
+            UNWIND $node_id_list AS node_id
+
+            // for each node id in every row, collect the random list of its neighbors.
+            CALL apoc.cypher.run(
+
+                'MATCH(cur_node) WHERE cur_node.ID = $node_id
+
+                // find the neighbors
+                MATCH (cur_node)--(neighbors)
+                WITH collect(neighbors.ID) AS neigh_ids
+                RETURN neigh_ids',
+                {{node_id: node_id}}) YIELD value
+
+            RETURN collect(value.neigh_ids) as neighbors
```

Comment:
We could model this more similar to `StellarGraph` itself by getting an edge list directly, instead of the pseudo adjacency matrix:

```suggestion
            MATCH (source)--(target)
            WHERE source.ID IN $node_id_list AND target.ID IN $node_id_list
            RETURN source.ID AS source, target.ID as target
```

Thoughts?

Reply:
Assuming that people using `Neo4jStellarGraph` are dealing with big graphs I went for the `O(edges incident to nodes)` approach over the `O(E)` approach,  but Neo4j might optimize your query to be `O(edges incident to nodes)` anyway - i'll try and profile to figure out.

Reply:
Ah, that's true. It'd be fine to file an issue for followup and leave this for later, because it's an implementation detail.

Reply:
i did some profiling with `EXPLAIN` and your query is optimized to `O(edges incident to nodes)`, not only that your query is ~10x faster! i guess from avoiding the unwind and list of lists 🤷‍♂️